spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindOverridableMethodCall.java

```diff
@@ -0,0 +1,274 @@
+import edu.umd.cs.findbugs.classfile.ClassDescriptor;
+import edu.umd.cs.findbugs.util.MultiMap;
+
+public class FindOverridableMethodCall extends OpcodeStackDetector {
+
+    private static class CallerInfo {
+        XMethod method;
+        SourceLineAnnotation sourceLine;
+
+        CallerInfo(XMethod m, SourceLineAnnotation sl) {
+            method = m;
+            sourceLine = sl;
+        }
+    }
+
+    private static final Map<XMethod, CallerInfo> callerConstructors = new HashMap<>();
+    private static final Map<XMethod, CallerInfo> callerClones = new HashMap<>();
+    private static final Map<XMethod, XMethod> callsToOverridable = new HashMap<>();
+    private static final MultiMap<XMethod, XMethod> callerToCalleeMap = new MultiMap<>(ArrayList.class);
+    private static final MultiMap<XMethod, XMethod> calleeToCallerMap = new MultiMap<>(ArrayList.class);
+
+
+    private final BugAccumulator bugAccumulator;
+
+    public FindOverridableMethodCall(BugReporter bugReporter) {
+        this.bugAccumulator = new BugAccumulator(bugReporter);
+    }
+
+    @Override
+    public void visit(JavaClass obj) {
+        super.visit(obj);
+        callerConstructors.clear();
+        callerClones.clear();
+        callsToOverridable.clear();
+        callerToCalleeMap.clear();
+        calleeToCallerMap.clear();
+    }
+
+    @Override
+    public void visitAfter(JavaClass obj) {
+        bugAccumulator.reportAccumulatedBugs();
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (seen == Const.INVOKEINTERFACE || seen == Const.INVOKEVIRTUAL) {
+            XMethod method = getXMethodOperand();
+            if (method == null) {
+                return;
+            }
+            OpcodeStack.Item item = stack.getStackItem(0);
+
+            if (item.getRegisterNumber() == 0 && Const.CONSTRUCTOR_NAME.equals(getMethodName())) {
+                if (checkDirectCase(method, "MC_OVERRIDABLE_METHOD_CALL_IN_CONSTRUCTOR", LOW_PRIORITY)) {
+                    checkAndRecordCallFromConstructor(getXMethod(), method);
+                }
+
+            } else if ("clone".equals(getMethodName())
+                    && (("()" + getClassDescriptor().getSignature()).equals(getMethodSig())
+                            || "()Ljava.lang.Object;".equals(getMethodSig()))
+                    && item.getReturnValueOf() != null
+                    && item.getReturnValueOf().equals(superClone(getXClass()))) {
+                if (checkDirectCase(method, "MC_OVERRIDABLE_METHOD_CALL_IN_CLONE", NORMAL_PRIORITY)) {
+                    checkAndRecordCallFromClone(getXMethod(), method);
+                }
+
+            } else if (item.getRegisterNumber() == 0
+                    && (getXMethod().isPrivate() || getXMethod().isFinal())) {
+                if (!method.isPrivate() && !method.isFinal()) {
+                    checkAndRecordCallToOverridable(getXMethod(), method);
+                } else if (method.isPrivate() || method.isFinal()) {
```

Comment:
This condition seems needless.

```suggestion
                } else {
```